File: tests/api/utils/api_client.py
import requests
from typing import Dict, Any


class APIClient:
    def __init__(self, base_url: str):
        self.base_url = base_url.rstrip("/")

    def _url(self, path: str) -> str:
        if not path.startswith("/"):
            path = "/" + path
        return self.base_url + path

    def get(self, path: str, headers: Dict[str, str] | None = None, params: Dict[str, Any] | None = None):
        return requests.get(self._url(path), headers=headers, params=params)

    def post(self, path: str, json: Dict[str, Any] | None = None, headers: Dict[str, str] | None = None, data=None):
        return requests.post(self._url(path), json=json, headers=headers, data=data)

    def put(self, path: str, json: Dict[str, Any] | None = None, headers: Dict[str, str] | None = None):
        return requests.put(self._url(path), json=json, headers=headers)

    def patch(self, path: str, json: Dict[str, Any] | None = None, headers: Dict[str, str] | None = None):
        return requests.patch(self._url(path), json=json, headers=headers)

    def delete(self, path: str, headers: Dict[str, str] | None = None):
        return requests.delete(self._url(path), headers=headers)

File: frontend/src/apiClient.js
const API_BASE_URL =
  import.meta.env.VITE_API_BASE_URL || "http://127.0.0.1:8000";

async function request(path, options = {}) {
  const url = `${API_BASE_URL}${path}`;
  const resp = await fetch(url, {
    headers: {
      "Content-Type": "application/json",
      ...(options.headers || {}),
    },
    ...options,
  });

  const text = await resp.text();
  let data;
  try {
    data = text ? JSON.parse(text) : null;
  } catch {
    data = text;
  }

  if (!resp.ok) {
    const error = new Error(
      data && data.detail ? data.detail : `Request failed: ${resp.status}`
    );
    error.status = resp.status;
    error.data = data;
    throw error;
  }

  return data;
}

export function apiGet(path, options = {}) {
  return request(path, { ...options, method: "GET" });
}

export function apiPost(path, body, options = {}) {
  return request(path, {
    ...options,
    method: "POST",
    body: JSON.stringify(body),
  });
}

export function apiPut(path, body, options = {}) {
  return request(path, {
    ...options,
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export function apiPatch(path, body, options = {}) {
  return request(path, {
    ...options,
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

export function apiDelete(path, options = {}) {
  return request(path, { ...options, method: "DELETE" });
}
